cmd/ical-notifier/notifier.go

```diff
@@ -0,0 +1,353 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"rss"
+	"time"
+
+	"gopkg.in/gomail.v2"
+
+	ics "github.com/arran4/golang-ical"
+	"github.com/jm-lemmi/ical-relay/compare"
+	"github.com/jm-lemmi/ical-relay/datastore"
+	"github.com/jm-lemmi/ical-relay/helpers"
+	log "github.com/sirupsen/logrus"
+)
+
+func notifyChanges(notifierName string, notifier datastore.Notifier) error {
+	requestLogger := log.WithFields(log.Fields{"notifier": notifierName})
+	requestLogger.Infoln("Running Notifier!")
+
+	// get source
+	currentICS, err := getSource(notifier.Source)
+	if err != nil {
+		log.Error("Failed to get source for notifier", notifierName, err)
+		return err
+	}
+	// compare to history on file
+	historyFilename := conf.General.StoragePath + "notifystore/" + notifierName + "-past.ics"
+	historyICS, err := helpers.LoadCalFile(historyFilename)
+
+	// if history file does not exist, create it
+	if err != nil {
+		log.Info("History file does not exist, saving for the first time")
+		helpers.WriteCalFile(currentICS, historyFilename)
+		return nil
+	}
+
+	added, deleted, changed_old, changed_new := compare.Compare(historyICS, currentICS)
+
+	if len(added)+len(deleted)+len(changed_old) == 0 {
+		log.Info("No changes detected.")
+		return nil
+	}
+	log.Debug("Changes detected: " + fmt.Sprint(len(added)) + " added, " + fmt.Sprint(len(deleted)) + " deleted, " + fmt.Sprint(len(changed_old)) + " changed")
+
+	//get notifier including recipients because the method to get all notifiers returns only the notifiers with an empty recipients slice
```

Comment:
shouldnt this be fixed in the original function instead of this hack?

Reply:
This is also currently broken. I still get an empty slice

Reply:
It did work for me with this fix.. Anyways, the underlying problem is the database datastore which retrieves the notifier excluding the recipients. We should either returns just a list of strings with the name of the notifier, or the Notifier struct including the recipients.

Reply:
im FileDataStore gibt GetNotifiers das komplette struct korrekt zurück. Das ist also ein Problem der DB

Reply:
diese Funktion hat ja auch gar keinen Zugriff auf den vollen Datastore, sondern bekommt ja schon den konkreten Notifier übergeben. Zugriff sollte auch nur über diese Variable dann gemacht werden.

Reply:
we might as well just change the bool value in the dbDatastore to true and remove the extra step here. I'd expect the function to also return the recipients unless its name says otherwise